docs/specs/saml.md

```diff
@@ -0,0 +1,697 @@
+
+### <a id="2_3"></a> The Logout Endpoint
+
+- The URL is https://example.authgear.cloud/saml2/logout/ENTITY_ID. Where `ENTITY_ID` is the `id` of the service provider as specified in the config `saml.service_providers[index].id`. As a result, each service provider will have a independent logout endpoint.
+- The metadata for each service provider will includes the above login endpoint.
+- When the endpoint receives a GET or POST request with a `<LogoutRequest>` provided, the [Single Logout flow](#2) will be triggered. A `<LogoutResponse>` will be returned. The binding used to return the response will be same as the binding used to receive the `<LogoutRequest>`.
+- When the endpoint receives a GET request without a `<LogoutRequest>` provided, and if a valid IdPSession exist, redirect the user to the `/logout` endpoint.
+- When the endpoint receives a GET request without a `<LogoutRequest>` provided, and if a valid IdPSession does not exist, redirect the user to the login endpoint as defined in [The Login Endpoint](#1_6).
+
+## <a id="3"></a> NameID
+
+### <a id="3_1"></a> Supported NameIDFormat
+
+- urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified
+  - See the below [Configure NameID when using urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified](#3_1_1) section.
+- urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress
+  - Use the `email` attribute of the user as the NameID. If the user does not have an `email` attribute, the login will fail.
+
+#### <a id="3_1_1"></a> Configure NameID when using urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified
+
+```yaml
+saml:
+  service_providers:
+    - id: ENTITY_ID
+      name_id_format: urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified
+      name_id_attribute_pointer: /username
+```
+
+- `name_id_attribute_pointer`: The JSON pointer pointing to a user's standard attribute or custom attribute which will be used as the NameID. The login will fail if the attribute does not exist or it is empty. The default is `/sub`, which is the user id.
+
+### <a id="3_2"> Configs
+
+```yaml
+saml:
+  service_providers:
+    - id: ENTITY_ID
+      name_id_format: urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified
+      name_id_attribute_pointer: /username
+```
+
+- `name_id_format`: Optional. The `NameIDFormat` used in SAML assertions. Default `urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified`.
+- `name_id_attribute_pointer`: Optional. See the above [Configure NameID when using urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified](#3_1_1) section for details. Default `/sub`. Only effective if `name_id_format` is `urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified`.
+
+### <a id="3_3"></a> Unsupported `NameIDFormat`
+
+The following `NameIDFormat`s are not supported at the moment, but they are commonly supported by different SAML IdPs so we would like to support it in the future:
+
+- urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
+- urn:oasis:names:tc:SAML:2.0:nameid-format:transient
```

Comment:
Document the support of them is backward compatible. In particular, how are we going to support `persistent` with `name_id_attribute_pointer` in a backward compatible way.

Reply:
I think `name_id_attribute_pointer` is only for `unspecified` type of NameID. Other NameIDFormat does not read the config.